urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/stationserver/FaultyStationServerImpl.java

```diff
@@ -0,0 +1,497 @@
+		try {
+			tcpProperties.setProperty(this.stationType.getStationCode(),
+					InetAddress.getLocalHost().getHostName() + " " + port);
+		} catch (UnknownHostException e) {
+			e.printStackTrace();
+		}
+		this.tcpServer = new TCPServer(Integer.valueOf(port));
+		this.log.debug(this.stationType.getStationCode()
+				+ ":TCPServer started on port[" + port + "]");
+		return this;
+	}
+
+	public void addOtherUDPStationHostNPort(StationType stationType,
+			String host, String port) {
+		udpProperties.setProperty(stationType.getStationCode(), host + " "
+				+ port);
+	}
+
+	public void addOtherTCPStationHostNPort(StationType stationType,
+			String host, String port) {
+		tcpProperties.setProperty(stationType.getStationCode(), host + " "
+				+ port);
+	}
+
+	/*
+	 * start's TCP server on the port number given in the properites file
+	 */
+	public FaultyStationServerImpl startTCPPServer() {
+		// read the host and port number to be started on for this station
+		String port = extractHostAndPortNumber((String) tcpProperties
+				.get(this.stationType.getStationCode()))[1];
+		return startTCPPServer(port);
+	}
+
+	/**
+	 * constructor
+	 * 
+	 * @param stationType
+	 */
+	public FaultyStationServerImpl(final StationType stationType) {
+		// initialize this station type
+		this.stationType = stationType;
+		// initialize the logger
+		this.log = Logger.getLogger(this.stationType.getStationCode());
+		log.info(stationType.getStationName() + " initialized!");
+		try {
+			// load the udp.properties and tcp.properties file, if not found
+			// write a new file
+			// with the default values
+			udpProperties.load(new FileInputStream("./udp.properties"));
+
+		} catch (FileNotFoundException e) {
+			log.error("config properties not found");
+			applyDefaultUDPSettings();
+		} catch (IOException e) {
+			e.printStackTrace();
+			log.error(e.getMessage());
+		}
+
+		try {
+			tcpProperties.load(new FileInputStream("tcp.properties"));
+		} catch (IOException e) {
+			e.printStackTrace();
+			log.error(e.getMessage());
+			applyDefaultTCPSettings();
+		}
+	}
+
+	// default values, for each station's port and host address
+
+	private void applyDefaultUDPSettings() {
+
+		udpProperties.setProperty("SPVM", "localhost 9091");
+		udpProperties.setProperty("SPB", "localhost 9092");
+		udpProperties.setProperty("SPL", "localhost 9093");
+
+		try {
+			udpProperties.store(new FileOutputStream("./udp.properties"), null);
+			log.debug(this.stationType.getStationCode()
+					+ ":Default config properites initialized");
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	// default values, for each station's port and host address
+
+	private void applyDefaultTCPSettings() {
+
+		tcpProperties.setProperty("SPVM", "localhost 9094");
+		tcpProperties.setProperty("SPB", "localhost 9095");
+		tcpProperties.setProperty("SPL", "localhost 9096");
+
+		try {
+			tcpProperties.store(new FileOutputStream("./tcp.properties"), null);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileOutputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).